slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+    if isinstance(ir, Transfer):
+        return f"transfer({encode_ir_for_compare(ir.call_value)})"
+    if isinstance(ir, Send):
+        return f"send({encode_ir_for_compare(ir.call_value)})"
+    if isinstance(ir, Unpack):  # TODO: improve
+        return "unpack"
+    if isinstance(ir, InitArray):  # TODO: improve
+        return "init_array"
+    if isinstance(ir, Function):  # TODO: investigate this
+        return "function_solc"
+
+    # variables
+    if isinstance(ir, Constant):
+        return f"constant({ntype(ir.type)})"
+    if isinstance(ir, SolidityVariableComposed):
+        return f"solidity_variable_composed({ir.name})"
+    if isinstance(ir, SolidityVariable):
+        return f"solidity_variable{ir.name}"
+    if isinstance(ir, TemporaryVariable):
+        return "temporary_variable"
+    if isinstance(ir, ReferenceVariable):
+        return f"reference({ntype(ir.type)})"
+    if isinstance(ir, LocalVariable):
+        return f"local_solc_variable({ir.location})"
+    if isinstance(ir, StateVariable):
+        return f"state_solc_variable({ntype(ir.type)})"
+    if isinstance(ir, LocalVariableInitFromTuple):
+        return "local_variable_init_tuple"
+    if isinstance(ir, TupleVariable):
+        return "tuple_variable"
+
+    # default
+    return ""
+
+
+def get_proxy_implementation_slot(proxy: Contract) -> Optional[SlotInfo]:
+    """
+    Gets information about the storage slot where a proxy's implementation address is stored.
+    Args:
+        proxy: A Contract object (proxy.is_upgradeable_proxy should be true).
+
+    Returns:
+        (`SlotInfo`) | None : A dictionary of the slot information.
+    """
+
+    delegate = get_proxy_implementation_var(proxy)
+    if isinstance(delegate, StateVariable):
+        if not delegate.is_constant and not delegate.is_immutable:
+            srs = SlitherReadStorage([proxy], 20)
+            return srs.get_storage_slot(delegate, proxy)
+        if delegate.is_constant and delegate.type.name == "bytes32":
+            return SlotInfo(
+                name=delegate.name,
+                type_string="address",
+                slot=int(delegate.expression.value, 16),
+                size=160,
+                offset=0,
+            )
+    return None
+
+
+def get_proxy_implementation_var(proxy: Contract) -> Optional[Variable]:
+    """
+    Gets the Variable that stores a proxy's implementation address. Uses data dependency to trace any LocalVariable
+    that is passed into a delegatecall as the target address back to its data source, ideally a StateVariable.
+    Args:
+        proxy: A Contract object (proxy.is_upgradeable_proxy should be true).
+
+    Returns:
+        (`Variable`) | None : The variable, ideally a StateVariable, which stores the proxy's implementation address.
+    """
+    available_functions = proxy.available_functions_as_dict()
+    if not proxy.is_upgradeable_proxy or not available_functions["fallback()"]:
```

Comment:
I think this can lead to issues if a function is named `fallback()`

We should iterate over the function, and look for `function. is_fallback`